Single inproc process died and stuck when restarting

I'm running a single Infrakit container with multiple plugins running via `infrakit plugin start manager group swarm terraform`. The process panic'd in the terraform plugin (I suspect that there was a network issue connecting to the NFS volume hosing the files being traversed):

```
$ docker logs infrakit.95d76cdnsf6608a31zukm493c.kso3q6ehgdf1h590arpstnsse
...
ibm_compute_vm_instance.instance-1506029775: Refreshing state... (ID: 40178701)
ibm_compute_vm_instance.instance-1506030036: Refreshing state... (ID: 40180467)
ibm_compute_vm_instance.instance-1506030016: Refreshing state... (ID: 40179481)
ibm_compute_vm_instance.instance-1506030017: Refreshing state... (ID: 40179483)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x1a17d88]

goroutine 112 [running]:
github.com/docker/infrakit/pkg/provider/terraform/instance.(*plugin).handleFiles.func1(0xc4201e5fe0, 0x1c, 0x0, 0x0, 0x2e87640, 0xc4202b3b00, 0x2e87640, 0xc4202b3b00)
	/go/src/github.com/docker/infrakit/pkg/provider/terraform/instance/apply.go:177 +0x58
github.com/docker/infrakit/vendor/github.com/spf13/afero.walk(0x2ea3480, 0x2f26448, 0xc42029aae0, 0xf, 0x2e9fdc0, 0xc42043fe10, 0xc4205a4780, 0x0, 0x4093b5)
	/go/src/github.com/docker/infrakit/vendor/github.com/spf13/afero/path.go:65 +0x32b
github.com/docker/infrakit/vendor/github.com/spf13/afero.Walk(0x2ea3480, 0x2f26448, 0xc42029aae0, 0xf, 0xc4205a4780, 0x1, 0xc4205a4780)
	/go/src/github.com/docker/infrakit/vendor/github.com/spf13/afero/path.go:107 +0x103
github.com/docker/infrakit/vendor/github.com/spf13/afero.Afero.Walk(0x2ea3480, 0x2f26448, 0xc42029aae0, 0xf, 0xc4205a4780, 0x0, 0x1b7d760)
	/go/src/github.com/docker/infrakit/vendor/github.com/spf13/afero/path.go:99 +0x53
github.com/docker/infrakit/pkg/provider/terraform/instance.(*plugin).handleFiles(0xc4201c6d20, 0xc4204fab40, 0xc420751f08, 0x0, 0x0)
	/go/src/github.com/docker/infrakit/pkg/provider/terraform/instance/apply.go:201 +0x26f
github.com/docker/infrakit/pkg/provider/terraform/instance.(*plugin).terraformApply.func1(0xc4201c6d20)
	/go/src/github.com/docker/infrakit/pkg/provider/terraform/instance/apply.go:64 +0xe5
created by github.com/docker/infrakit/pkg/provider/terraform/instance.(*plugin).terraformApply
	/go/src/github.com/docker/infrakit/pkg/provider/terraform/instance/apply.go:87 +0x23e
$
```

This caused the container to go into an `Exited` status:
```
f363ed3ffb66        infrakit:srk   "infrakit plugin -..."   16 hours ago        Exited (2) 14 hours ago                                  infrakit.95d76cdnsf6608a31zukm493c.kso3q6ehgdf1h590arpstnsse
```

Infrakit is running as a `docker service` and was restarted, the replacement container simply has:
```
$ docker logs infrakit.95d76cdnsf6608a31zukm493c.uyrw0y0dibmubkmhfny77c57d
WARN[09-22|02:27:42] Creating directory                       module=cli/local dir=/root/.infrakit/cli fn=github.com/docker/infrakit/pkg/cli/local.Setup
INFO[09-22|02:27:42] Launching                                module=cli/plugin kind=manager name=group fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func2
INFO[09-22|02:27:42] Launching                                module=cli/plugin kind=group name=group-stateless fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func2
INFO[09-22|02:27:42] Launching                                module=cli/plugin kind=swarm name=swarm fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func2
INFO[09-22|02:27:42] Launching                                module=cli/plugin kind=terraform name=terraform fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func2
INFO[09-22|02:27:42] Done waiting on plugin starts            module=cli/plugin fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func2
$
```

I suspect that the issue is that the plugin sockets and pid files weren't cleaned up from the crashed container:
```
$ docker exec infrakit.95d76cdnsf6608a31zukm493c.uyrw0y0dibmubkmhfny77c57d ls -l /infrakit/plugins
total 20
-rw-r--r--    1 root     root             1 Sep 21 21:36 24864.pid
srwxr-xr-x    1 root     root             0 Sep 21 21:36 group
srwxr-xr-x    1 root     root             0 Sep 21 21:36 group-stateless
-rw-r--r--    1 root     root             1 Sep 21 21:36 group-stateless.pid
-rw-r--r--    1 root     root             1 Sep 21 21:36 group.pid
srwxr-xr-x    1 root     root             0 Sep 21 21:36 swarm
-rw-r--r--    1 root     root             1 Sep 21 21:36 swarm.pid
srwxr-xr-x    1 root     root             0 Sep 21 21:36 terraform
-rw-r--r--    1 root     root             1 Sep 21 21:36 terraform.pid
```
```
$ docker exec infrakit.95d76cdnsf6608a31zukm493c.uyrw0y0dibmubkmhfny77c57d infrakit plugin ls
WARN[09-22|13:37:55] cannot get types for this kind           module=cli/plugin err="Post http://h: dial unix /infrakit/plugins/group: connect: connection refused" addr=/infrakit/plugins/group fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func1
WARN[09-22|13:37:55] cannot get types for this kind           module=cli/plugin err="Post http://h: dial unix /infrakit/plugins/group-stateless: connect: connection refused" addr=/infrakit/plugins/group-stateless fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func1
WARN[09-22|13:37:55] cannot get types for this kind           module=cli/plugin err="Post http://h: dial unix /infrakit/plugins/swarm: connect: connection refused" addr=/infrakit/plugins/swarm fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func1
WARN[09-22|13:37:55] cannot get types for this kind           module=cli/plugin err="Post http://h: dial unix /infrakit/plugins/terraform: connect: connection refused" addr=/infrakit/plugins/terraform fn=github.com/docker/infrakit/cmd/infrakit/plugin.Command.func1
INTERFACE           NAME                          LISTEN
                    group                         /infrakit/plugins/group
                    group-stateless               /infrakit/plugins/group-stateless
                    swarm                         /infrakit/plugins/swarm
                    terraform                     /infrakit/plugins/terraform
```

If the inproc process crashes how will the plugin files get cleaned up so that a replacement container can successfully be started?

--- a/cmd/infrakit/plugin/plugin.go
+++ b/cmd/infrakit/plugin/plugin.go
@@ -139,7 +139,7 @@ func Command(plugins func() discovery.Plugins) *cobra.Command {
 	}
 
 	configURL := start.Flags().String("config-url", "", "URL for the startup configs")
-	mustAll := start.Flags().Bool("all", true, "Panic if any plugin fails to start")
+	mustAll := start.Flags().Bool("all", false, "Panic if any plugin fails to start")
 	templateFlags, toJSON, _, processTemplate := base.TemplateProcessor(plugins)
 	start.Flags().AddFlagSet(templateFlags)
 
@@ -175,7 +175,15 @@ func Command(plugins func() discovery.Plugins) *cobra.Command {
 		if err != nil {
 			return err
 		}
-		defer pluginManager.Stop()
+		defer func() {
+			if r := recover(); r != nil {
+				log.Error("Error occurred. Recovered but exiting.", "err", r)
+				pluginManager.TerminateRunning()
+			}
+			pluginManager.WaitForAllShutdown()
+			log.Info("All plugins shutdown")
+			pluginManager.Stop()
+		}()
 
 		if len(args) == 0 {
 
@@ -227,9 +235,6 @@ func Command(plugins func() discovery.Plugins) *cobra.Command {
 		pluginManager.WaitStarting()
 
 		log.Info("Done waiting on plugin starts")
-
-		pluginManager.WaitForAllShutdown()
-		log.Info("All plugins shutdown")
 
 		return nil
 	}

--- a/pkg/cli/v0/event/cmd.go
+++ b/pkg/cli/v0/event/cmd.go
@@ -15,7 +15,8 @@ var log = logutil.New("module", "cli/v1/event")
 func init() {
 	cli.Register(event.InterfaceSpec,
 		[]cli.CmdBuilder{
-			Event,
+			Ls,
+			Tail,
 		})
 }
 

--- a/pkg/cli/v0/event/ls.go
+++ b/pkg/cli/v0/event/ls.go
@@ -14,8 +14,8 @@ import (
 func Ls(name string, services *cli.Services) *cobra.Command {
 
 	ls := &cobra.Command{
-		Use:   "ls",
-		Short: "List event",
+		Use:   "topics",
+		Short: "List event topics",
 	}
 
 	long := ls.Flags().BoolP("long", "l", false, "Print full path")

--- a/pkg/cli/v0/metadata/cmd.go
+++ b/pkg/cli/v0/metadata/cmd.go
@@ -13,13 +13,24 @@ import (
 var log = logutil.New("module", "cli/v1/metadata")
 
 func init() {
+	// cli.Register(metadata.InterfaceSpec,
+	// 	[]cli.CmdBuilder{
+	// 		Metadata,
+	// 	})
+	// cli.Register(metadata.UpdatableInterfaceSpec,
+	// 	[]cli.CmdBuilder{
+	// 		Updatable,
+	// 	})
 	cli.Register(metadata.InterfaceSpec,
 		[]cli.CmdBuilder{
-			Metadata,
+			Ls,
+			Cat,
 		})
 	cli.Register(metadata.UpdatableInterfaceSpec,
 		[]cli.CmdBuilder{
-			Updatable,
+			Ls,
+			Cat,
+			Change,
 		})
 }
 

--- a/pkg/cli/v0/metadata/ls.go
+++ b/pkg/cli/v0/metadata/ls.go
@@ -15,7 +15,7 @@ import (
 func Ls(name string, services *cli.Services) *cobra.Command {
 
 	ls := &cobra.Command{
-		Use:   "ls",
+		Use:   "vars",
 		Short: "List metadata",
 	}
 

--- a/pkg/run/manager/manager.go
+++ b/pkg/run/manager/manager.go
@@ -32,6 +32,7 @@ func ManagePlugins(rules []launch.Rule,
 		plugins:      plugins,
 		mustAll:      mustAll,
 		scanInterval: scanInterval,
+		running:      []string{},
 	}
 	return m, m.Start(rules)
 }
@@ -51,11 +52,18 @@ type Manager struct {
 	wgStartAll  sync.WaitGroup
 	started     chan plugin.Name
 	lock        sync.RWMutex
+
+	running []string // lookups of those started
 }
 
 // Rules returns a list of plugins that can be launched via this manager
 func (m *Manager) Rules() []launch.Rule {
 	return m.rules
+}
+
+// TerminateRunning terminates those that have been started.
+func (m *Manager) TerminateRunning() error {
+	return m.Terminate(m.running)
 }
 
 // Terminate stops the plugins.  Note this is accomplished by sending a signal TERM to the
@@ -227,6 +235,8 @@ func (m *Manager) WaitForAllShutdown() {
 		case target := <-m.started:
 
 			lookup, _ := target.GetLookupAndType()
+
+			m.running = append(m.running, lookup)
 
 			if _, has := seen[lookup]; !has {
 				log.Debug("Start watching", "lookup", lookup)

